src/core.js

```diff
@@ -27,8 +31,51 @@ extend( QUnit, {
 			delete testEnvironment.teardown;
 		}
 
-		config.modules.push( currentModule );
-		config.currentModule = currentModule;
+		module = createModule();
+
+		moduleFns = {
+			beforeEach: setHook( module, "beforeEach" ),
+			afterEach: setHook( module, "afterEach" )
+		};
+
+		if ( executeNow instanceof Function ) {
+			config.moduleStack.push( module );
+			setCurrentModule( module );
+			executeNow.call( module.testEnvironment, moduleFns );
+			config.moduleStack.pop();
+			module = module.parentModule || currentModule;
+		}
+
+		setCurrentModule( module );
+
+		function createModule() {
+			var parentModule = config.moduleStack.length ?
+				config.moduleStack.slice( -1 )[ 0 ] : null;
+			var moduleName = parentModule !== null ?
+				[ parentModule.name, name ].join( " > " ) : name;
+			var module = {
+				name: moduleName,
+				parentModule: parentModule,
+				tests: []
+			};
+
+			var env = {};
+			if ( parentModule ) {
+				extend( env, parentModule.testEnvironment );
```

Comment:
Arbitrary inheritance is possible in ES3, too:
```js
var createObject = (function( Base ) {
	return function( prototype ) {
		Base.prototype = prototype;
		return new Base();
	};
})( Function() );
…
env = createObject( parentModule.testEnvironment );
```

The above is similar to a stripped-down version of lodash's [`baseCreate`](https://github.com/lodash/lodash/blob/d407d8aa08a6f739441a82e3b7c9d8a5cf66ba16/lodash.js#L1987-L2005), but you could easily go all the way to [`_.create`](https://lodash.com/docs#create) if you want this part to look like `module.testEnvironment = createObject( parentModule.testEnvironment, testEnvironment )` (removing the need for a separate `extend`).